core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -257,31 +449,84 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-
-      if (quota.equals(defaultQuota))
-        this.overriddenQuota.remove(clientId)
-      else
-        this.overriddenQuota.put(clientId, quota)
+      val quotaId = QuotaId.quotaId(sanitizedUser, clientId)
+      val userInfo = sanitizedUser match {
+        case Some(ConfigEntityName.Default) => "default user "
+        case Some(user) => "user " + user + " "
+        case None => ""
+      }
+      val clientIdInfo = clientId match {
+        case Some(ConfigEntityName.Default) => "default client-id"
+        case Some(id) => "client-id " + id
+        case None => ""
+      }
+      quota match {
+        case Some(newQuota) =>
+          logger.info(s"Changing ${apiKey} quota for ${userInfo}${clientIdInfo} to ${newQuota.bound}")
+          overriddenQuota.put(quotaId, newQuota)
+          (sanitizedUser, clientId) match {
+            case (Some(u), Some(c)) => quotaTypesEnabled |= QuotaTypes.UserClientIdQuotaEnabled
+            case (Some(u), None) => quotaTypesEnabled |= QuotaTypes.UserQuotaEnabled
+            case (None, Some(c)) => quotaTypesEnabled |= QuotaTypes.ClientIdQuotaEnabled
+            case (None, None) =>
+          }
+        case None =>
+          logger.info(s"Removing ${apiKey} quota for ${userInfo}${clientIdInfo}")
+          overriddenQuota.remove(quotaId)
+      }
 
-      // Change the underlying metric config if the sensor has been created
+      val quotaMetricName = clientRateMetricName(sanitizedUser.getOrElse(""), clientId.getOrElse(""))
       val allMetrics = metrics.metrics()
-      val quotaMetricName = clientRateMetricName(clientId)
-      if (allMetrics.containsKey(quotaMetricName)) {
-        logger.info(s"Sensor for clientId $clientId already exists. Changing quota to ${quota.bound()} in MetricConfig")
-        allMetrics.get(quotaMetricName).config(getQuotaMetricConfig(quota))
+
+      // If multiple-levels of quotas are defined or if this is a default quota update, traverse metrics
+      // to find all affected values. Otherwise, update just the single matching one.
+      val singleUpdate = quotaTypesEnabled match {
+        case QuotaTypes.NoQuotas | QuotaTypes.ClientIdQuotaEnabled | QuotaTypes.UserQuotaEnabled | QuotaTypes.UserClientIdQuotaEnabled =>
+          !sanitizedUser.filter(_ == ConfigEntityName.Default).isDefined && !clientId.filter(_ == ConfigEntityName.Default).isDefined
+        case _ => false
       }
+      if (singleUpdate) {
+          // Change the underlying metric config if the sensor has been created
+          if (allMetrics.containsKey(quotaMetricName)) {
+            val metricConfigEntity = quotaEntity(sanitizedUser.getOrElse(""), clientId.getOrElse(""))
+            val newQuota = metricConfigEntity.quota
+            logger.info(s"Sensor for ${userInfo}${clientIdInfo} already exists. Changing quota to ${newQuota.bound()} in MetricConfig")
+            allMetrics.get(quotaMetricName).config(getQuotaMetricConfig(newQuota))
```

Comment:
This is an existing corner case issue. It's possible for a metric to get expired between the time the check in line 490 and here. So, it's safer to save the result of allMetrics.get(quotaMetricName) to a local val and only update the config if it's not null.